File: frontend_new/orchestrator/context_summarizer.py
"""
Smart Context Summarization System for ATLAS
Implements progressive context summarization to manage token limits efficiently
"""

import json
import logging
import os
from typing import Dict, List, Optional, Tuple
from datetime import datetime

logger = logging.getLogger(__name__)

class ContextSummarizer:
    """
    Handles intelligent context summarization for long conversations
    """
    
    def __init__(self, max_context_tokens: int = 45000, summary_ratio: float = 0.3):
        self.max_context_tokens = max_context_tokens
        self.summary_ratio = summary_ratio  # 30% of original context
        self.conversation_state = {
            "current_session": [],
            "summarized_history": "",
            "last_summary_timestamp": None,
            "token_count_estimate": 0
        }
    
    def estimate_tokens(self, text: str) -> int:
        """
        Rough token estimation: ~4 chars per token for most languages
        More accurate for Ukrainian/Russian: ~3.5 chars per token
        """
        return len(text) // 4 if text else 0
    
    def should_summarize(self, new_content: str) -> bool:
        """
        Determine if we need to summarize context
        """
        estimated_tokens = (
            self.conversation_state["token_count_estimate"] + 
            self.estimate_tokens(new_content)
        )
        
        return estimated_tokens > self.max_context_tokens
    
    def create_summary_prompt(self, context_to_summarize: List[Dict]) -> str:
        """
        Create a focused summary prompt for the AI
        """
        context_text = ""
        for item in context_to_summarize:
            if isinstance(item, dict):
                role = item.get("role", "unknown")
                content = item.get("content", "")
                context_text += f"\n{role.upper()}: {content}\n"
            else:
                context_text += f"\n{item}\n"
        
        summary_prompt = f"""
Створи стислий, але інформативний підсумок наступної розмови, зберігаючи:
1. Ключові технічні деталі та рішення
2. Важливі помилки та їх виправлення  
3. Стан системи та налаштування
4. Основні досягнення та результати

Контекст для підсумовування:
{context_text}

Підсумок повинен бути структурованим та займати не більше {int(len(context_text) * self.summary_ratio)} символів.
"""
        return summary_prompt
    
    def summarize_context(self, context_to_summarize: List[Dict], ai_client) -> str:
        """
        Generate summary using AI
        """
        try:
            summary_prompt = self.create_summary_prompt(context_to_summarize)
            
            response = ai_client.generate_response(
                messages=[{
                    "role": "system", 
                    "content": "Ти експерт з підсумовування технічних розмов. Створи точний та стислий підсумок."
                }, {
                    "role": "user", 
                    "content": summary_prompt
                }],
                max_tokens=2000
            )
            
            summary = response.get("content", "")
            logger.info(f"Context summarized: {len(''.join([str(item) for item in context_to_summarize]))} → {len(summary)} chars")
            
            return summary
            
        except Exception as e:
            logger.error(f"Failed to create summary: {e}")
            # Fallback: simple truncation summary
            return self._create_fallback_summary(context_to_summarize)
    
    def _create_fallback_summary(self, context: List[Dict]) -> str:
        """
        Simple fallback summary when AI summarization fails
        """
        summary_parts = []
        
        for item in context[-5:]:  # Last 5 interactions
            if isinstance(item, dict):
                role = item.get("role", "")
                content = item.get("content", "")[:200]  # First 200 chars
                summary_parts.append(f"{role}: {content}...")
        
        return f"[АВТОПІДСУМОК] Останні взаємодії:\n" + "\n".join(summary_parts)
    
    def process_new_interaction(self, user_input: str, ai_response: str, ai_client=None) -> Dict:
        """
        Process new interaction and return optimized context
        """
        # Add new interaction
        new_interaction = [
            {"role": "user", "content": user_input},
            {"role": "assistant", "content": ai_response}
        ]
        
        self.conversation_state["current_session"].extend(new_interaction)
        
        # Check if we need to summarize
        if self.should_summarize(user_input + ai_response):
            if ai_client and self.conversation_state["current_session"]:
                # Create summary of older context
                context_to_summarize = self.conversation_state["current_session"][:-4]  # All except last 2 interactions
                
                if context_to_summarize:
                    new_summary = self.summarize_context(context_to_summarize, ai_client)
                    
                    # Update conversation state
                    self.conversation_state["summarized_history"] += f"\n\n[ПІДСУМОК {datetime.now().strftime('%H:%M')}]\n{new_summary}"
                    self.conversation_state["current_session"] = self.conversation_state["current_session"][-4:]  # Keep last 2 interactions
                    self.conversation_state["last_summary_timestamp"] = datetime.now().isoformat()
        
        # Update token estimate
        self.conversation_state["token_count_estimate"] = (
            self.estimate_tokens(self.conversation_state["summarized_history"]) +
            self.estimate_tokens(''.join([str(item) for item in self.conversation_state["current_session"]]))
        )
        
        return self.get_optimized_context()
    
    def get_optimized_context(self) -> Dict:
        """
        Return optimized context ready for AI
        """
        context = {
            "summarized_history": self.conversation_state["summarized_history"],
            "current_session": self.conversation_state["current_session"],
            "estimated_tokens": self.conversation_state["token_count_estimate"],
            "last_summary": self.conversation_state["last_summary_timestamp"]
        }
        
        return context
    
    def format_for_ai_prompt(self) -> str:
        """
        Format context for inclusion in AI prompt
        """
        formatted_context = ""
        
        if self.conversation_state["summarized_history"]:
            formatted_context += f"[ПОПЕРЕДНІЙ КОНТЕКСТ]\n{self.conversation_state['summarized_history']}\n\n"
        
        formatted_context += "[ПОТОЧНА СЕСІЯ]\n"
        for item in self.conversation_state["current_session"]:
            if isinstance(item, dict):
                role = item.get("role", "").upper()
                content = item.get("content", "")
                formatted_context += f"{role}: {content}\n\n"
        
        return formatted_context
    
    def save_state(self, filepath: str):
        """Save conversation state to file"""
        try:
            with open(filepath, 'w', encoding='utf-8') as f:
                json.dump(self.conversation_state, f, ensure_ascii=False, indent=2)
            logger.info(f"Context state saved to {filepath}")
        except Exception as e:
            logger.error(f"Failed to save context state: {e}")
    
    def load_state(self, filepath: str):
        """Load conversation state from file"""
        try:
            if os.path.exists(filepath):
                with open(filepath, 'r', encoding='utf-8') as f:
                    self.conversation_state = json.load(f)
                logger.info(f"Context state loaded from {filepath}")
                return True
        except Exception as e:
            logger.error(f"Failed to load context state: {e}")
        return False

# Utility functions for easy integration
def create_smart_context_manager(max_tokens=45000):
    """Create a context manager instance"""
    return ContextSummarizer(max_context_tokens=max_tokens)

def integrate_with_orchestrator(context_manager, orchestrator_instance):
    """
    Integration helper for existing orchestrator
    """
    # This will be implemented based on your orchestrator structure
    pass

File: frontend_new/orchestrator/context_summarizer.js
/**
 * Smart Context Summarization System for ATLAS
 * Implements progressive context summarization to manage token limits efficiently
 */

import fs from 'fs';
import path from 'path';

class ContextSummarizer {
    constructor(maxContextTokens = 45000, summaryRatio = 0.3) {
        this.maxContextTokens = maxContextTokens;
        this.summaryRatio = summaryRatio; // 30% of original context
        this.conversationState = {
            currentSession: [],
            summarizedHistory: "",
            lastSummaryTimestamp: null,
            tokenCountEstimate: 0
        };
        
        // State file path
        this.stateFile = path.resolve(process.cwd(), '../logs/context_state.json');
        
        // Load existing state if available
        this.loadState();
    }
    
    /**
     * Rough token estimation: ~4 chars per token for most languages
     * More accurate for Ukrainian/Russian: ~3.5 chars per token
     */
    estimateTokens(text) {
        return text ? Math.ceil(text.length / 4) : 0;
    }
    
    /**
     * Determine if we need to summarize context
     */
    shouldSummarize(newContent) {
        const estimatedTokens = this.conversationState.tokenCountEstimate + this.estimateTokens(newContent);
        return estimatedTokens > this.maxContextTokens;
    }
    
    /**
     * Create a focused summary prompt for the AI
     */
    createSummaryPrompt(contextToSummarize) {
        let contextText = "";
        for (const item of contextToSummarize) {
            if (typeof item === 'object' && item !== null) {
                const role = item.role || "unknown";
                const content = item.content || "";
                contextText += `\n${role.toUpperCase()}: ${content}\n`;
            } else {
                contextText += `\n${item}\n`;
            }
        }
        
        const targetLength = Math.floor(contextText.length * this.summaryRatio);
        
        return `
Створи стислий, але інформативний підсумок наступної розмови, зберігаючи:
1. Ключові технічні деталі та рішення
2. Важливі помилки та їх виправлення  
3. Стан системи та налаштування
4. Основні досягнення та результати

Контекст для підсумовування:
${contextText}

Підсумок повинен бути структурованим та займати не більше ${targetLength} символів.
`;
    }
    
    /**
     * Generate summary using AI client
     */
    async summarizeContext(contextToSummarize, aiClient) {
        try {
            const summaryPrompt = this.createSummaryPrompt(contextToSummarize);
            
            const response = await aiClient.generateResponse([{
                role: "system", 
                content: "Ти експерт з підсумовування технічних розмов. Створи точний та стислий підсумок."
            }, {
                role: "user", 
                content: summaryPrompt
            }], {
                maxTokens: 2000,
                temperature: 0.3
            });
            
            const summary = response.content || "";
            const originalLength = contextToSummarize.map(item => JSON.stringify(item)).join('').length;
            
            console.log(`[ContextSummarizer] Context summarized: ${originalLength} → ${summary.length} chars`);
            
            return summary;
            
        } catch (error) {
            console.error(`[ContextSummarizer] Failed to create summary:`, error);
            return this.createFallbackSummary(contextToSummarize);
        }
    }
    
    /**
     * Simple fallback summary when AI summarization fails
     */
    createFallbackSummary(context) {
        const summaryParts = [];
        
        // Take last 5 interactions
        const recentContext = context.slice(-5);
        
        for (const item of recentContext) {
            if (typeof item === 'object' && item !== null) {
                const role = item.role || "";
                const content = (item.content || "").substring(0, 200); // First 200 chars
                summaryParts.push(`${role}: ${content}...`);
            }
        }
        
        return `[АВТОПІДСУМОК] Останні взаємодії:\n${summaryParts.join('\n')}`;
    }
    
    /**
     * Process new interaction and return optimized context
     */
    async processNewInteraction(userInput, aiResponse, aiClient = null) {
        // Add new interaction
        const newInteraction = [
            { role: "user", content: userInput },
            { role: "assistant", content: aiResponse }
        ];
        
        this.conversationState.currentSession.push(...newInteraction);
        
        // Check if we need to summarize
        if (this.shouldSummarize(userInput + aiResponse)) {
            if (aiClient && this.conversationState.currentSession.length > 4) {
                // Keep last 2 interactions, summarize the rest
                const contextToSummarize = this.conversationState.currentSession.slice(0, -4);
                
                if (contextToSummarize.length > 0) {
                    const newSummary = await this.summarizeContext(contextToSummarize, aiClient);
                    
                    // Update conversation state
                    const timestamp = new Date().toISOString();
                    this.conversationState.summarizedHistory += `\n\n[ПІДСУМОК ${new Date().toLocaleTimeString('uk-UA', { hour: '2-digit', minute: '2-digit' })}]\n${newSummary}`;
                    this.conversationState.currentSession = this.conversationState.currentSession.slice(-4); // Keep last 2 interactions
                    this.conversationState.lastSummaryTimestamp = timestamp;
                    
                    console.log(`[ContextSummarizer] Created summary at ${timestamp}, current session reduced to ${this.conversationState.currentSession.length} items`);
                }
            }
        }
        
        // Update token estimate
        this.conversationState.tokenCountEstimate = 
            this.estimateTokens(this.conversationState.summarizedHistory) +
            this.estimateTokens(JSON.stringify(this.conversationState.currentSession));
        
        // Save state
        this.saveState();
        
        return this.getOptimizedContext();
    }
    
    /**
     * Return optimized context ready for AI
     */
    getOptimizedContext() {
        return {
            summarizedHistory: this.conversationState.summarizedHistory,
            currentSession: this.conversationState.currentSession,
            estimatedTokens: this.conversationState.tokenCountEstimate,
            lastSummary: this.conversationState.lastSummaryTimestamp
        };
    }
    
    /**
     * Format context for inclusion in AI prompt
     */
    formatForAiPrompt() {
        let formattedContext = "";
        
        if (this.conversationState.summarizedHistory) {
            formattedContext += `[ПОПЕРЕДНІЙ КОНТЕКСТ]\n${this.conversationState.summarizedHistory}\n\n`;
        }
        
        if (this.conversationState.currentSession.length > 0) {
            formattedContext += "[ПОТОЧНА СЕСІЯ]\n";
            for (const item of this.conversationState.currentSession) {
                if (typeof item === 'object' && item !== null) {
                    const role = (item.role || "").toUpperCase();
                    const content = item.content || "";
                    formattedContext += `${role}: ${content}\n\n`;
                }
            }
        }
        
        return formattedContext;
    }
    
    /**
     * Save conversation state to file
     */
    saveState() {
        try {
            // Ensure logs directory exists
            const logsDir = path.dirname(this.stateFile);
            if (!fs.existsSync(logsDir)) {
                fs.mkdirSync(logsDir, { recursive: true });
            }
            
            fs.writeFileSync(this.stateFile, JSON.stringify(this.conversationState, null, 2), 'utf8');
            console.log(`[ContextSummarizer] State saved to ${this.stateFile}`);
        } catch (error) {
            console.error(`[ContextSummarizer] Failed to save state:`, error);
        }
    }
    
    /**
     * Load conversation state from file
     */
    loadState() {
        try {
            if (fs.existsSync(this.stateFile)) {
                const stateData = fs.readFileSync(this.stateFile, 'utf8');
                this.conversationState = JSON.parse(stateData);
                console.log(`[ContextSummarizer] State loaded from ${this.stateFile}`);
                return true;
            }
        } catch (error) {
            console.error(`[ContextSummarizer] Failed to load state:`, error);
        }
        return false;
    }
    
    /**
     * Clear conversation state (start fresh)
     */
    clearState() {
        this.conversationState = {
            currentSession: [],
            summarizedHistory: "",
            lastSummaryTimestamp: null,
            tokenCountEstimate: 0
        };
        this.saveState();
        console.log('[ContextSummarizer] State cleared');
    }
    
    /**
     * Get context statistics
     */
    getStats() {
        return {
            currentSessionLength: this.conversationState.currentSession.length,
            summarizedHistoryLength: this.conversationState.summarizedHistory.length,
            estimatedTokens: this.conversationState.tokenCountEstimate,
            lastSummary: this.conversationState.lastSummaryTimestamp
        };
    }
}

// Utility functions
export function createSmartContextManager(maxTokens = 45000) {
    return new ContextSummarizer(maxTokens);
}

export { ContextSummarizer };
export default ContextSummarizer;
